patches/minecraft/net/minecraft/block/KelpTopBlock.java.patch

```diff
@@ -0,0 +1,15 @@
+--- a/net/minecraft/block/KelpTopBlock.java
++++ b/net/minecraft/block/KelpTopBlock.java
+@@ -57,9 +57,12 @@
+       } else {
+          BlockPos blockpos = p_196267_3_.func_177984_a();
+          BlockState blockstate = p_196267_2_.func_180495_p(blockpos);
++         if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, true)) {
```

Comment:
This is not how the crop growth events are supposed to be implemented. You are supposed to replace the existing if-check where the random value is, passing the result of the random check to the event, eg. line 60 should look like this:

    if (blockstate.getBlock() == Blocks.WATER && state.get(AGE) < 25 &&
        ForgeHooks.onCropsGrowPre(p_196267_2_, blockpos, p_196267_1_, random.nextDouble() < 0.14D))
    { ... }

The whole point is to be able to *force* the growth even if the vanilla logic says that the block does not grow *and* prevent it when vanilla logic says it does. Currently your patch only covers the latter.

Reply:
Ok, I have updated the PR